File: 222.count-complete-tree-nodes.java
/*
 * @lc app=leetcode id=222 lang=java
 *
 * [222] Count Complete Tree Nodes
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {

        int res = 0;
        int treeHeight = countHeight(root);

        while (root != null) {
            if (countHeight(root.right) == treeHeight - 1) {
                res += 1 << treeHeight;
                root = root.right;
            } else {
                res += 1 << treeHeight - 1;
                root = root.left;
            }

            treeHeight--;
        }

        return res;
        
    }

    private int countHeight(TreeNode root) {
        
        if (root == null) {
            return -1;
        }

        return 1 + countHeight(root.left);
    }
}
// @lc code=end

File: 222.count-complete-tree-nodes.py
#
# @lc app=leetcode id=222 lang=python3
#
# [222] Count Complete Tree Nodes
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        res = 0
        tree_height = self.countHeight(root)

        while root:
            if self.countHeight(root.right) == tree_height - 1:
                res += 1 << tree_height
                root = root.right
            else:
                res += 1 << tree_height - 1
                root = root.left
            
            tree_height -= 1
        
        return res

    def countHeight(self, root):
        if not root:
            return -1
        else:
            return 1 + self.countHeight(root.left)
        
# @lc code=end

